src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceParameterService.cs

```diff
@@ -0,0 +1,153 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable disable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.AddParameter;
+using Microsoft.CodeAnalysis.CodeGeneration;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.IntroduceVariable;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace Microsoft.CodeAnalysis.CSharp.IntroduceVariable
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp), Shared]
+    internal partial class CSharpIntroduceParameterService : AbstractIntroduceParameterService<CSharpIntroduceParameterService, ExpressionSyntax, MethodDeclarationSyntax, InvocationExpressionSyntax>
+    {
+        [ImportingConstructor]
+        [SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification = "Used in test code: https://github.com/dotnet/roslyn/issues/42814")]
+        public CSharpIntroduceParameterService()
+        {
+        }
+
+        protected override async Task<Solution> IntroduceParameterAsync(SemanticDocument document, ExpressionSyntax expression, bool allOccurrences, bool trampoline, CancellationToken cancellationToken)
+        {
+            return await AbstractIntroduceParameterAsync(document, expression, allOccurrences, trampoline, cancellationToken).ConfigureAwait(false);
+        }
+
+        protected override async Task<Solution> RewriteCallSitesAsync(ExpressionSyntax expression, ImmutableDictionary<Document, List<InvocationExpressionSyntax>> callSites,
+         IMethodSymbol methodSymbol, CancellationToken cancellationToken)
+        {
+            var mappingDictionary = TieExpressionToParameters(expression, methodSymbol);
+            expression = expression.TrackNodes(mappingDictionary.Values);
+            var identifiers = expression.DescendantNodes().Where(node => node is IdentifierNameSyntax);
+
+            if (!callSites.Keys.Any())
+            {
+                return null;
+            }
+
+            var firstCallSite = callSites.Keys.First();
+
+            var currentSolution = firstCallSite.Project.Solution;
+            foreach (var keyValuePair in callSites)
```

Comment:
we should group all callsites by document, then process each documetn at a time.

Reply:
They already are `ImmutableDictionary<Document, List<InvocationExpressionSyntax>> callSites`

Reply:
true!

Reply:
however, we should group by project so that we process all docs in a project at the same time.  this will rpevent bouncing around many differnet compilations causing churn.

Reply:
Changed